### PR TITLE
Draining bug fix

### LINKED ISSUE #4731: Ember Server returns lots of errors where Blaze… doesn't?
Tested on **main**. Steps are as follows…

Add the following route definition to `EmberServerSimpleExample`:

```scala

        case req @ POST -> Root / "wut" =>
          Ok {
            req.body
              .through(fs2.text.utf8Decode[F])
              .map { str =>
                var c = 0
                var i = 0
                while (i < str.length) {
                  if (str.charAt(i) == 'a') {
                    c += 1
                  }
                  i += 1
                }

                s"$c\n"
              }
              .through(fs2.text.utf8Encode[F])
          }
```

Then, using `hey`, hit the server a lot with a file post. I used `scala-library-2.13.5.jar`, but you could like… do anything, I think, so long as it's reasonably large. Example invocation:

```
$ hey -D target/bg-jobs/sbt_8f92a902/target/511dd676/11210362/scala-library-2.13.5.jar -m POST http://127.0.0.1:8080/wut
```

About half of the two hundred responses will be errors. The distribution looks like the following:

```
Status code distribution:
  [200]	103 responses

Error distribution:
  [1]	Post "http://127.0.0.1:8080/wut": read tcp 127.0.0.1:60679->127.0.0.1:8080: read: connection reset by peer
  [1]	Post "http://127.0.0.1:8080/wut": read tcp 127.0.0.1:60686->127.0.0.1:8080: read: connection reset by peer
  [1]	Post "http://127.0.0.1:8080/wut": read tcp 127.0.0.1:60687->127.0.0.1:8080: read: connection reset by peer
  [1]	Post "http://127.0.0.1:8080/wut": read tcp 127.0.0.1:60690->127.0.0.1:8080: read: connection reset by peer
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60680->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60681->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60682->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60683->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60684->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60685->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60688->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60689->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60691->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60692->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60693->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60694->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60695->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60696->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60697->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60698->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60699->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60700->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60701->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60702->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60703->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60704->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60705->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60706->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60707->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60708->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60709->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60711->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60712->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60713->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60714->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60715->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60716->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60717->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60718->127.0.0.1:8080: write: protocol wrong type for socket
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60719->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60720->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60721->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60722->127.0.0.1:8080: use of closed network connection
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60723->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60724->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60725->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60726->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60727->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60728->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60729->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60730->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60731->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60732->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60733->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60734->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60735->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60736->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60737->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60738->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60739->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60740->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60741->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60742->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60743->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60744->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60745->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60746->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60747->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60748->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60749->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60751->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60752->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60753->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60754->127.0.0.1:8080: write: protocol wrong type for socket
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60755->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60756->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60757->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60758->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60759->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60760->127.0.0.1:8080: write: protocol wrong type for socket
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60762->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60763->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60764->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60765->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60766->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60768->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60769->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60770->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60771->127.0.0.1:8080: write: protocol wrong type for socket
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60772->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60773->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60774->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60775->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60776->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60777->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60778->127.0.0.1:8080: write: broken pipe
  [1]	Post "http://127.0.0.1:8080/wut": write tcp 127.0.0.1:60779->127.0.0.1:8080: write: protocol wrong type for socket
```

If you do the same test with blaze-server (by editing `ExampleService`), you'll see 100% HTTP 200 responses.

### PR DESCRIPTION
**Pull request by RaasAhsan**
Fixes #4731. The issue here was that we would drain a socket's read buffer _before_ the response was written to the socket. This can manifest in the aforementioned bug because the input stream will only be evaluated when we try to write. But if we drain the connection before we write, we'll end up closing because we see unread bytes.

So the fix is to drain _after_ the response is written to the socket. I've also done some QoL to make things easier to read and follow. There might be some merge conflicts with #4735 which I can deal with

TODO:
- [x] Tests (Daniel's reproduction is a good one)
- [x] Run some more tests locally for this

### CODE REVIEW

File: ember-core/src/main/scala/org/http4s/ember/core/EmptyStreamError.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.ember.core
+
+private[ember] final case class EmptyStreamError() extends Exception("Cannot Parse Empty Stream")
```
Comment (rossabaker) [inline ember-core/src/main/scala/org/http4s/ember/core/EmptyStreamError.scala:19]: Nit: I don't love exceptions being case classes.  It implies equality where there is none.  I see you're using it as an extractor, but I usually catch these by type.

**Review by rossabaker — APPROVED** [review]
I won't die on the case class hill.

File: ember-core/src/main/scala/org/http4s/ember/core/EmptyStreamError.scala (hunk 1 above)
Comment (RaasAhsan) [reply ember-core/src/main/scala/org/http4s/ember/core/EmptyStreamError.scala:19]: This is also not a hill I'll die on, but every other exception type defined in that package is also a case class, and most of them are not used in a match clause :) . If we want to go one way or the other, I'm all for for following well-defined guidelines

Comment (rossabaker) [reply ember-core/src/main/scala/org/http4s/ember/core/EmptyStreamError.scala:19]: Agree, I care about consistency more than which way we go.  Let's merge, and we can revisit the case class bikeshed.

**Pull request merged**